substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java

```diff
@@ -690,51 +691,15 @@ private boolean runPointsToAnalysis(String imageName, OptionValues options, Debu
                 DuringAnalysisAccessImpl config = new DuringAnalysisAccessImpl(featureHandler, loader, bb, nativeLibraries, debug);
-                int numIterations = 0;
-                while (true) {
-                    try (Indent indent2 = debug.logAndIndent("new analysis iteration")) {
-                        /*
-                         * Do the analysis (which itself is done in a similar iterative process)
-                         */
-                        boolean analysisChanged = bb.finish();
-
-                        numIterations++;
-                        if (numIterations > 1000) {
-                            /*
-                             * Usually there are < 10 iterations. If we have so many iterations, we
-                             * probably have an endless loop (but at least we have a performance
-                             * problem because we re-start the analysis so often).
-                             */
-                            throw UserError.abort("Static analysis did not reach a fix point after %d iterations because a Feature keeps requesting new analysis iterations. " +
-                                            "The analysis itself %s find a change in type states in the last iteration.",
-                                            numIterations, analysisChanged ? "DID" : "DID NOT");
-                        }
-
-                        /*
-                         * Allow features to change the universe.
-                         */
-                        try (StopTimer t2 = bb.getProcessFeaturesTimer().start()) {
-                            int numTypes = aUniverse.getTypes().size();
-                            int numMethods = aUniverse.getMethods().size();
-                            int numFields = aUniverse.getFields().size();
-
+                String msg = ((NativeImagePointsToAnalysis) bb).iterativelyAnalyze(debug,
+                        (aUniverse) -> {
                             bb.getHostVM().notifyClassReachabilityListener(aUniverse, config);
                             featureHandler.forEachFeature(feature -> feature.duringAnalysis(config));
-
-                            if (!config.getAndResetRequireAnalysisIteration()) {
-                                if (numTypes != aUniverse.getTypes().size() || numMethods != aUniverse.getMethods().size() || numFields != aUniverse.getFields().size()) {
-                                    throw UserError.abort(
-                                                    "When a feature makes more types, methods, or fields reachable, it must require another analysis iteration via DuringAnalysisAccess.requireAnalysisIteration()");
-                                }
-                                break;
-                            }
-                        }
-                    }
+                            return !config.getAndResetRequireAnalysisIteration();
+                        });
+                if (msg.length() > 0) {
```

Comment:
I don't like this approach of returning a String on error. Instead you could just throw an `com.oracle.graal.pointsto.util.AnalysisError` from `PoinstToAnalysis.runAnalysis()` then catch it here and wrap it in a user error, something like: `throw UserError.abort(original, "Analysis step failed. Reason: %s.", original.getMessage());`

Reply:
fixed as suggested